mono/utils/mono-error.h

```diff
@@ -56,6 +56,38 @@ typedef struct _MonoError {
 	void *hidden_1 [12]; /*DON'T TOUCH */
 } MonoError;
 
+// FIXME This is duplicated.
+#define ERROR_DECL \
+	MonoError error_value; \
+	MonoError * const error = &error_value; \
+	error_init_internal (error);
+
+#define ERROR_DECL_NAME(x) \
+	MonoError x; \
+	error_init_internal (&x);
+
+// MonoError.hidden_0 is MonoErrorInternal.flags.
+// error_init is only now for compatibility at API boundaries.
+// Internal users should use ERROR_DECL.
+
+#undef error_init
+#define error_init(error) ((error)->error_code = (error)->hidden_0 = 0)
+#undef error_init
+#define error_init(error) // FIXME remove them all
+
+#undef error_init
+#define error_init(x) error_init_internal (x) // FIXME
+
```

Comment:
1. I had trouble getting it to compiler and took a shortcut of duplication. We should be able to clean it up further. This is very many steps forward, very few steps backwards. We can take a few steps forward sooner or later, but I wanted to get something mostly there.

2. Because first I had removed the cast, and then embarked on the large change. The first two lines can go away.

3. Casting is worse. The macro can take a pointer to anything and mess with it. By removing the casts, it is much safer.